buildpacks/metrics-agent/src/layers/install_agentmon.rs

```diff
@@ -0,0 +1,329 @@
+                Ok(ExistingLayerStrategy::Recreate)
+            }
+        } else {
+            self.section
+                .say_with_details("Clearing cache", "No url found in metadata");
+
+            Ok(ExistingLayerStrategy::Recreate)
+        }
+    }
+
+    fn migrate_incompatible_metadata(
+        &self,
+        _context: &libcnb::build::BuildContext<Self::Buildpack>,
+        _metadata: &GenericMetadata,
+    ) -> Result<
+        libcnb::layer::MetadataMigration<Self::Metadata>,
+        <Self::Buildpack as libcnb::Buildpack>::Error,
+    > {
+        self.section
+            .say_with_details("Clearing cache", "invalid metadata");
+
+        Ok(libcnb::layer::MetadataMigration::RecreateLayer)
+    }
+}
+
+fn write_execd(agentmon_path: &Path, layer_path: &Path) -> Result<PathBuf, DownloadAgentmonError> {
+    let agentmon_path = agentmon_path
+        .canonicalize()
+        .map_err(DownloadAgentmonError::CouldNotOpenFile)?;
+    let agentmon_path = agentmon_path.display();
+
+    // This script boots and runs agentmon in a loop
+    let agentmon_loop = {
+        let script = layer_path.join("agentmon_loop");
+
+        // Copy compiled binary from `bin/agentmon_loop.rs` to layer
+        fs_err::copy(additional_buildpack_binary_path!("agentmon_loop"), &script)
+            .map_err(DownloadAgentmonError::CouldNotWriteDestinationFile)?;
+
+        script
+            .canonicalize()
+            .map_err(DownloadAgentmonError::CouldNotOpenFile)?
+    };
+
+    // We use the exec.d to boot a process. This script MUST exit though as otherwise
+    // The container would never boot. To handle this we intentionally leak a process
+    let execd_script = {
+        let execd_script = layer_path.join("agentmon_exec.d");
+        let log_file = layer_path.join("background.log");
+        fs_err::write(&log_file, "")
+            .map_err(DownloadAgentmonError::CouldNotCreateDestinationFile)?;
+
+        let log_file = log_file.display();
+        let agentmon_loop = agentmon_loop.display();
+        write_bash_script(
+            &execd_script,
+            format!(
+                r#"
+                if [ -z "$AGENTMON_DEBUG" ]
+                then
+                    start-stop-daemon --start --background \
+                        --exec "{agentmon_loop}" \
+                        --output {log_file} \
+                        -- --path {agentmon_path}
+                else
+                    echo "To enable logging run with AGENTMON_DEBUG=1" >> {log_file}
+
+                    start-stop-daemon --start --background \
+                        --exec "{agentmon_loop}" \
+                        -- --path {agentmon_path}
+                fi
+            "#
+            ),
+        )
+        .map_err(DownloadAgentmonError::CouldNotWriteDestinationFile)?;
+
+        execd_script
+    };
+
+    Ok(execd_script)
+}
+
+#[cached]
+fn get_latest_url() -> Result<Url, GetUrlError> {
```

Comment:
We want to move to a more secure way of dealing with downloaded binaries for language runtimes and others. This implementation is not only excessively complex (cached macro, IO, URL parsing), it also doesn't verify he integrity of the download binary. 

The strategy implemented here also means that the behaviour at runtime could change or outright break, even when we don't release an updated buildpack. If a new agentmon version comes out, CI isn't run for it and customers immediately get the new version for their next build - including during moratoriums, etc.

I suggest that we add buildpack metadata for the URL and checksum of the most recent agentmon release which we can then use for downloading and verification. We can also add a simple automation that updates the `buildpack.toml` with the most recent version as a GHA that opens a PR for it.

This way we:
- Verify the binary, removing an attack vector
- Only explicitly update the agentmon version
- Enforce testing when a new agentmon version comes out
- Remove IO and therefore potential of errors at runtime
- Remove the need to memoize this function is a macro

Reply:
This logic was a straight 1:1 port of the existing buildpack logic. The easiest thing to move forward would be to hardcode the path. Agentmon hasn't been updated in 6 years https://github.com/heroku/agentmon.

I would love some standardization on some of this stuff. Ideally being able to have some code like:

```rust
libcnb::MaifestDownload::new("agentmon.manifest")
  .with_tag(ManifestTag::KeyValue("stack", "heroku-22")
  .with_tag(ManifestTag::Flag("latest")
  .validate(ManifestValidate::Sha, ManifestTag::Key("sha"))
  .tar_gz()
  .download(layer_path)
  .map_err(...)
```

Reply:
Updated to hardcode the value. Take a look, please.

Reply:
Updated with a SHA 256 checksum using libherokubuildpack. Please take another look